### ISSUE TITLE
In-tree build is not working (inconsistency in registered CommandLine or core dumped)

### ISSUE THREAD

**Issue by TheErk** [include-what-you-use/include-what-you-use#499]
When building IWYU in tree with either clang 4.0.1 or clang 5.0, the compilation succeed but I do end up with an unusable executable:

```
$ include-what-you-use
: CommandLine Error: Option 'asan-instrument-assembly' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```

I seek the issue a little bit and found some LLVM bug reference (https://reviews.llvm.org/rL240104) speaking of double link to some LLVM static lib.
I did examine the link line for IWYU and discover a lot of duplicates.
In fact all LLVM libs where duplicated on the command line.

I commented out the 
```
 target_link_libraries(include-what-you-use
    LLVMX86AsmParser # MC, MCParser, Support, X86CodeGen, X86Desc, X86Info
...
    LLVMDemangle
    )
```

in IWYU CMakeLists.txt and I end up with an apparently working iwyu:
```
$ include-what-you-use
error: no input files
error: unable to handle compilation, expected exactly one compiler job in ''
$ include-what-you-use --version
include-what-you-use 0.9 (git:576e80f) based on clang version 5.0.0 (tags/RELEASE_500/final)

```
However on a "real" call using compilation database generated by CMake on my project
(which correctly builds with either clang 4.0.1 or clang 5.0.0) I get core dumped.

My LLVM+Clang+IWYU build uses CMake + Ninja with the following parameters:

MY_LIBS=clang
MY_VERSION=5.0
INSTALL_DIR=<pathto>/local/clang
cmake .. -G Ninja \
         -DCLANG_ENABLE_BOOTSTRAP=On \
         -DCMAKE_BUILD_TYPE=MinSizeRel \
         -DCMAKE_CXX_FLAGS="-Wl,-rpath=${INSTALL_DIR}/lib/" \
         -DLLVM_TARGETS_TO_BUILD="X86" \
         -DLLVM_ENABLE_CXX1Y=On \
         -DLLVM_LINK_LLVM_DYLIB=On \
         -DLIBOMP_TSAN_SUPPORT=On \
         -DCMAKE_INSTALL_PREFIX:PATH=${INSTALL_DIR}/${MY_LIBS}/${MY_VERSION}/ 

What should I do to get a working IWYU in-tree build?

**Comment by kimgr** [comment]
I have a branch over at https://github.com/kimgr/include-what-you-use/tree/re-cmake to use a more official way to build IWYU with LLVM. It basically amounts to an out-of-tree build. I only got it half-finished, but I managed to build various configurations. There's some discussion on this cfe-dev thread: http://lists.llvm.org/pipermail/cfe-dev/2017-October/055669.html (including sketchy usage instructions).

Let me know if this works out better, and I'll have more motivation for putting more work into it.

**Comment by TheErk** [comment]
Thank you Kim, I'll try the branch and keep you posted.

**Comment by TheErk** [comment]
I did give a try and here is the outcome.
My testing process was the following:
Get LLVM 5.0 source set from tarballs: http://releases.llvm.org/download.html#5.0.0
My LLVM build includes: llvm, clang, lld, lldb, clang-tools-extra, openmp, compiler-rt.
I build those with:
```
$ cmake .. -G Ninja \
         -DCLANG_ENABLE_BOOTSTRAP=On \
         -DCMAKE_BUILD_TYPE=MinSizeRel \
         -DLLVM_TARGETS_TO_BUILD="X86" \
         -DLLVM_ENABLE_CXX1Y=On \
         -DLLVM_LINK_LLVM_DYLIB=On \
         -DLIBOMP_TSAN_SUPPORT=On \
         -DLLDB_DISABLE_PYTHON=On \ # because Swig versions 3.0.9 and 3.0.10 are incompatible with lldb.
         -DCMAKE_INSTALL_PREFIX:PATH=/home/me/local/clang

$ ninja
$ ninja install
```
Then I checked out your re-cmake branch.
I did setup:
```
set(IWYU_LLVM_CMAKE_DIR "${CMAKE_INSTALL_PREFIX}/lib/cmake/llvm" CACHE FILEPATH "Path to lib/cmake/llvm directory")
set(IWYU_CLANG_CMAKE_DIR "${CMAKE_INSTALL_PREFIX}/lib/cmake/clang" CACHE FILEPATH "Path to lib/cmake/llvm directory")
```
First of all I tried to point those to LLVM build dir but If I do that I end up with iwyu linked against LLVM-5.0.0.so located in the LLVM build tree (even after I installed iwyu in the same install prefix as LLVM).
When I switched to install location the reference is OK, but the iwyu executable ***in the build tree*** is not usable because the libLLVM-5.0.so => not found. This is probably due to the fact that LLVM macro is playing with RPATH and $ORIGIN, as shown by:
```
$ objdump -p b/include-what-you-use | grep PATH
RUNPATH              $ORIGIN/../lib
```
With this setup I end up with the very same issue as reported:
```
$ /home/me/local/clang/bin/include-what-you-use
: CommandLine Error: Option 'asan-instrument-assembly' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```
Looking at the link line generated by the build system I saw several duplicates of static LLVM/clang libraries.
Here is the initial link line:

`/usr/bin/c++   -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -w -ffunction-sections -fdata-sections  -Wl,-rpath-link,  -Wl,-O3 -Wl,--gc-sections CMakeFiles/include-what-you-use.dir/iwyu.cc.o CMakeFiles/include-what-you-use.dir/iwyu_ast_util.cc.o CMakeFiles/include-what-you-use.dir/iwyu_cache.cc.o CMakeFiles/include-what-you-use.dir/iwyu_driver.cc.o CMakeFiles/include-what-you-use.dir/iwyu_getopt.cc.o CMakeFiles/include-what-you-use.dir/iwyu_globals.cc.o CMakeFiles/include-what-you-use.dir/iwyu_include_picker.cc.o CMakeFiles/include-what-you-use.dir/iwyu_lexer_utils.cc.o CMakeFiles/include-what-you-use.dir/iwyu_location_util.cc.o CMakeFiles/include-what-you-use.dir/iwyu_output.cc.o CMakeFiles/include-what-you-use.dir/iwyu_path_util.cc.o CMakeFiles/include-what-you-use.dir/iwyu_preprocessor.cc.o CMakeFiles/include-what-you-use.dir/iwyu_verrs.cc.o  -o include-what-you-use -L/home/me/local/clang/lib -Wl,-rpath,"\$ORIGIN/../lib" -lpthread /home/me/local/clang/lib/libLLVMSupport.a /home/me/local/clang/lib/libLLVMX86AsmParser.a /home/me/local/clang/lib/libLLVMX86Desc.a /home/me/local/clang/lib/libLLVMX86Info.a /home/me/local/clang/lib/libclangBasic.a /home/me/local/clang/lib/libclangLex.a /home/me/local/clang/lib/libclangAST.a /home/me/local/clang/lib/libclangSema.a /home/me/local/clang/lib/libclangFrontend.a /home/me/local/clang/lib/libclangDriver.a /home/me/local/clang/lib/libLLVMMCDisassembler.a /home/me/local/clang/lib/libLLVMObject.a /home/me/local/clang/lib/libLLVMBitReader.a /home/me/local/clang/lib/libLLVMX86AsmPrinter.a /home/me/local/clang/lib/libLLVMX86Utils.a /home/me/local/clang/lib/libLLVMCore.a /home/me/local/clang/lib/libLLVMBinaryFormat.a /home/me/local/clang/lib/libLLVMMCParser.a /home/me/local/clang/lib/libLLVMMC.a /home/me/local/clang/lib/libLLVMSupport.a -lrt -ldl -ltinfo -lpthread -lz -lm /home/me/local/clang/lib/libLLVMDemangle.a /home/me/local/clang/lib/libclangParse.a /home/me/local/clang/lib/libclangSerialization.a /home/me/local/clang/lib/libclangSema.a /home/me/local/clang/lib/libclangAnalysis.a /home/me/local/clang/lib/libclangEdit.a /home/me/local/clang/lib/libclangAST.a /home/me/local/clang/lib/libclangLex.a /home/me/local/clang/lib/libclangBasic.a /home/me/local/clang/lib/libLLVM-5.0.so`

I did try to selectively remove libs from "target_link_libraries(include-what-you-use ..."

The one finally working (without the "more than once" error) is very simple:
```
target_link_libraries(include-what-you-use
  clangFrontend
  clangDriver
  )
```
With that build I do have a working iwyu (with no core dump).
The generated link line is:

`/usr/bin/c++   -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -w -ffunction-sections -fdata-sections  -Wl,-rpath-link,  -Wl,-O3 -Wl,--gc-sections CMakeFiles/include-what-you-use.dir/iwyu.cc.o CMakeFiles/include-what-you-use.dir/iwyu_ast_util.cc.o CMakeFiles/include-what-you-use.dir/iwyu_cache.cc.o CMakeFiles/include-what-you-use.dir/iwyu_driver.cc.o CMakeFiles/include-what-you-use.dir/iwyu_getopt.cc.o CMakeFiles/include-what-you-use.dir/iwyu_globals.cc.o CMakeFiles/include-what-you-use.dir/iwyu_include_picker.cc.o CMakeFiles/include-what-you-use.dir/iwyu_lexer_utils.cc.o CMakeFiles/include-what-you-use.dir/iwyu_location_util.cc.o CMakeFiles/include-what-you-use.dir/iwyu_output.cc.o CMakeFiles/include-what-you-use.dir/iwyu_path_util.cc.o CMakeFiles/include-what-you-use.dir/iwyu_preprocessor.cc.o CMakeFiles/include-what-you-use.dir/iwyu_verrs.cc.o  -o include-what-you-use -L/home/me/local/clang/lib -Wl,-rpath,"\$ORIGIN/../lib" -lpthread /home/me/local/clang/lib/libclangFrontend.a /home/me/local/clang/lib/libclangDriver.a /home/me/local/clang/lib/libclangParse.a /home/me/local/clang/lib/libclangSerialization.a /home/me/local/clang/lib/libclangSema.a /home/me/local/clang/lib/libclangEdit.a /home/me/local/clang/lib/libclangAnalysis.a /home/me/local/clang/lib/libclangAST.a /home/me/local/clang/lib/libclangLex.a /home/me/local/clang/lib/libclangBasic.a /home/me/local/clang/lib/libLLVM-5.0.so`

This is weird. I don't know why I need to do that because IWYU source code faithfully use the removed llvm/clang libs.  

All my builds (llvm or iwyu) are done with the host (Debian Stretch) gcc (version 6.3.0 20170516 (Debian 6.3.0-18)) + host ld.
I may try with clang and/or lld from the clang build but could you tell me your setup ?

Which compiler do you use to compile IWYU?
Do you build clang with BUILD_SHARED_LIBS and so gets shared libs instead of static ones?

Could you post here the link line which leads to a working IWYU for you?

I don't quite understand why too much libs on the link line may lead to non-working iwyu...
or may be I should use lld which does more clever things than ld.

**Comment by TheErk** [comment]
My current (working) tree is there:
https://github.com/TheErk/include-what-you-use/tree/re-cmake

I did
1) fork iwyu main repo
2) merge in kimr/re-cmake
3) commit my changes
4) push

 May be it will be easier to discuss with a forked repo.

**Comment by kimgr** [comment]
I'm a little short on time, but this is probably the main standout for me:

> Do you build clang with BUILD_SHARED_LIBS and so gets shared libs instead of static ones?

I've never tried any shared lib configuration with LLVM, and I can imagine there are dragons there.

**Comment by TheErk** [comment]
I'll look into this but in our llvm/clang build we set LLVM_LINK_LLVM_DYLIB=On in order to lower the size of llvm/clang binaries. Otherwise other tool/lib are static.

**Comment by kimgr** [comment]
I think the problem you're seeing is because your setup both statically and dynamically links most of LLVM, hence the "more than once" error. Somehow you end up with two copies of `ClAsanInstrumentAssembly` in your executable (I think it's highly suspicious that LLVM has a command-line options declared in a low-level library, but that's probably the only construct that makes any noise about it -- the same thing could happen silently for any symbol.)

I intuitively followed LWYU (link-what-you-use) when I hacked up that CMake configuration; I didn't want IWYU's build to rely on implicit link dependencies from Clang. I think the duplicated `.a` link deps are actually benign, I think the real problem is that we get both static (libLLVM*.a) and dynamic (libLLVM-5.0.so) linking of LLVM.

Can you try putting the duplicated `clang*.a` dependencies back, but leave out the LLVM stuff, and see if that works?

**Comment by TheErk** [comment]
You may be right.
I'll try that.
I won't be able to try before next Monday though.

Thanks for examining this.

**Comment by kimgr** [comment]
FWIW, here's my expanded link command:

```
/usr/bin/c++  -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -fno-common -Woverloaded-virtual -fno-strict-aliasing -fno-rtti -std=c++11 -g  -Wl,-allow-shlib-undefined tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_ast_util.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_cache.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_driver.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_getopt.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_globals.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_include_picker.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_lexer_utils.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_location_util.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_output.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_path_util.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_preprocessor.cc.o tools/clang/tools/include-what-you-use/CMakeFiles/include-what-you-use.dir/iwyu_verrs.cc.o  -o bin/include-what-you-use  lib/libclangFrontend.a lib/libclangSerialization.a lib/libclangDriver.a lib/libclangParse.a lib/libclangSema.a lib/libclangAnalysis.a lib/libclangAST.a lib/libclangBasic.a lib/libclangEdit.a lib/libclangLex.a lib/libLLVMX86AsmParser.a lib/libLLVMX86CodeGen.a lib/libLLVMX86Desc.a lib/libLLVMX86AsmPrinter.a lib/libLLVMX86Info.a lib/libLLVMX86Utils.a lib/libLLVMCodeGen.a lib/libLLVMipo.a lib/libLLVMScalarOpts.a lib/libLLVMInstCombine.a lib/libLLVMTransformUtils.a lib/libLLVMTarget.a lib/libLLVMAnalysis.a lib/libLLVMOption.a lib/libLLVMMCDisassembler.a lib/libLLVMMCParser.a lib/libLLVMMC.a lib/libLLVMProfileData.a lib/libLLVMObject.a lib/libLLVMBitReader.a lib/libLLVMCore.a lib/libLLVMBinaryFormat.a lib/libLLVMSupport.a lib/libLLVMDemangle.a -lpthread -lz -ldl lib/libclangAST.a lib/libclangLex.a lib/libclangBasic.a lib/libLLVMAsmPrinter.a lib/libLLVMDebugInfoCodeView.a lib/libLLVMDebugInfoMSF.a lib/libLLVMGlobalISel.a lib/libLLVMSelectionDAG.a lib/libLLVMCodeGen.a lib/libLLVMTarget.a lib/libLLVMScalarOpts.a lib/libLLVMInstCombine.a lib/libLLVMBitWriter.a lib/libLLVMIRReader.a lib/libLLVMAsmParser.a lib/libLLVMInstrumentation.a lib/libLLVMLinker.a lib/libLLVMVectorize.a lib/libLLVMTransformUtils.a lib/libLLVMAnalysis.a lib/libLLVMProfileData.a lib/libLLVMObject.a lib/libLLVMMCParser.a lib/libLLVMMC.a lib/libLLVMBitReader.a lib/libLLVMCore.a lib/libLLVMBinaryFormat.a lib/libLLVMSupport.a lib/libLLVMDemangle.a -lrt -lpthread -lm -ldl
```

**Comment by TheErk** [comment]
Leaving out only LLVM related link seems to work.
I'm re-compiling LLVM with static lib in order to see it works as well.
Note that other LLVM-based tool have similar issue;
https://cmake.org/pipermail/cmake/2017-April/065349.html

**Comment by TheErk** [comment]
You were right about multiple linking to LLVM.so and/or LLVMXXX.a libs.

I think I found a robust way to do the link in either case, have a look at this commit:
https://github.com/TheErk/include-what-you-use/commit/6a26f91a718f06835b2293cae17857cc2ec265d6

I checked whether if LLVM (shared lib) target exist in order to link to it otherwise link to static libs.
This works for me with LLVM+Clang compiled with LLVM.so shared lib or with the bunch of static libs.

I you are ok with it I can rewrite a clean branch and submiit a merge request.

**Comment by kimgr** [comment]
Thanks, this looks compelling!

Two concerns:

1) That CMakeLists.txt over in my branch is just a draft, and I'm not sure I'm entirely happy with having to provide two paths to LLVM and Clang CMake module dirs. I'll keep pinging the Clang lists about whether this is the best we can do.

2) The CMake discussion thread you linked to (thanks for that, very enlightening!) mentioned this:

> LLVM can be
> - A bunch of libLLVM*.a files
> - A bunch of libLLVM*.so files
> - A single libLLVM.so file

It looks like your change only covers the first and third case, or at least the comment makes me believe it only covers for the case with a single .so. But I guess if the LLVM libraries are individually shared, that's covered by the else-leg also, right?

I'm not familiar with the `TARGET` keyword in CMake, I'll have to read up on that and think some more. But overall this looks really enticing!

**Comment by TheErk** [comment]
Concerning 1. yes right having to provide path to LLVM/Clang module is not nice.
Usually one tries to find them at "usual" places then bail out requiring the user to provide some hint.
I'll try to provide such an example in my branch.

Concerning 2. the current CMakeLists.txt should cover all 3 cases already because when you link
to say LLVMSupport this is the name of the CMake 'target' which can either be static or shared depending on the LLVM build configuration.

Concerning `TARGET` keyword, I can explain a little more.
A CMake `TARGET` is the name provided to `add_executable`, `add_library` or `add_custom_target`.
See https://cmake.org/cmake/help/v3.7/manual/cmake-buildsystem.7.html

In our case the LLVM `target` is only defined iff `LLVM_LINK_LLVM_DYLIB` has been set to `On` when building LLVM. This setup effectively defines a new `target`.

To be complete when building out-of-tree (i.e. building iwyu outside LLVM/Clang tree) those targets are imported targets (https://cmake.org/cmake/help/v3.7/manual/cmake-buildsystem.7.html#imported-targets) which are defined by the LLVM, Clang modules exported and installed with LLVM/Clang.

The current setup does not work [yet] when building in-tree because when adding iwyu as a subdir
of llvm/tools/clang/tools  it comes before the definition of the LLVM target which is
defined in  llvm/tools/llvm-shlib.

I'll try to fix that too. I'll have a look to clang main CMakeLists.txt which is supposed to do that (detect when building in-tree or not).

**Comment by TheErk** [comment]
Kim,

Have a look at the tip of https://github.com/TheErk/include-what-you-use/tree/re-cmake.
Commit 4e1f9494358cebc741d811f643378abb4a96b0a3
should make IWYU compile clean in any case:

in- or out-of LLVM/Clang source tree.
with ou without LLVM_LINK_LLVM_DYLIB being On.

**Comment by kimgr** [comment]
Cool, thanks! I wonder if it would be nicer to pull out all the LLVM/Clang discovery into a separate .cmake file and include it? It might even be possible to expose a flag from there to say e.g. `IWYU_LINK_WITH_LLVM_SHLIB` which makes it more straightforward to select `LLVM` vs `LLVMSupport` & friends.

**Comment by TheErk** [comment]
It is possible.
It would make the CMakeLists.txt more readable at first sight.

Concerning `IWYU_LINK_WITH_LLVM_SHLIB` I don't known. I guess I would rather let IWYU build system adapt itself to the way LLVM/Clang were built and pick the appropriate lib.

That said the extra including CMake file could provide a macro which computes the best
set of libs IWYU wants to link to.

iwyu_get_llvm_clang_libs(FROM_LLVM 
                                            LLVMX86AsmParser LLVMX86Desc LLVMX86Info
                                         FROM_CLANG 
                                             clangBasic clangLex clangAST
                                             clangSema clangFrontend clangDriver
                                         NEEDED IWYU_LIBS)

then you 
target_link_libraries(include-what-you-use ${IWYU_LIBS})

may be it is worth asking to Clang/LLVM guys how to do that?
There should be some implicit rules they apply for that shared vs non-shared case.

**Comment by kimgr** [comment]
I don't think I want to get too clever, or I'll never be able to understand it two months from now :)

But separating out the LLVM/Clang probing would be nice, I think. It would be useful if the LLVM crowd could provide that, even.

**Comment by TheErk** [comment]
I totally agree with you.

**Comment by kimgr** [comment]
I've been playing around with this a little today.

I can't get it to work consistently on my Ubuntu box -- it looks like the Debian packaging doesn't include the CMake modules for Clang. Or something. I have `libclang-X.Y-dev` packages of 5 different versions, but only clang-4.0's CMake files under /usr/share/llvm-4.0/cmake.

I'm also a little concerned about Windows -- I no longer have a Windows system to test this on.

Since Clang seems to be the oddball, maybe we should consider something like:

```diff
-  if(EXISTS ${IWYU_LLVM_CMAKE_DIR} AND EXISTS ${IWYU_CLANG_CMAKE_DIR})
-    message(STATUS "IWYU: LLVM CMake module dir = ${IWYU_LLVM_CMAKE_DIR}")
-    message(STATUS "IWYU: Clang CMake module dir = ${IWYU_CLANG_CMAKE_DIR}")
-  else()
+  if(NOT EXISTS ${IWYU_LLVM_CMAKE_DIR})
     message(FATAL_ERROR
-      "IWYU: Don't know how to find LLVM and/or Clang CMake modules. "
-      "Use -DIWYU_CLANG_CMAKE_DIR=/xyz -DIWYU_LLVM_CMAKE_DIR=/xyz")
+      "IWYU: Don't know how to find LLVM CMake modules. "
+      "Use -DIWYU_LLVM_CMAKE_DIR=/xyz")
   endif()
 
+  if(NOT EXISTS ${IWYU_CLANG_CMAKE_DIR})
+    message(FATAL_ERROR
+      "IWYU: Don't know how to find CLANG CMake modules. "
+      "Use -DIWYU_CLANG_CMAKE_DIR=/xyz")
+  endif()
+
+  message(STATUS "IWYU: LLVM CMake module dir = ${IWYU_LLVM_CMAKE_DIR}")
+  message(STATUS "IWYU: Clang CMake module dir = ${IWYU_CLANG_CMAKE_DIR}")
```

That way, if probing for LLVM works, all we need to provide is a matching Clang dir. If only there was a way to install it :)

**Comment by TheErk** [comment]
Hi Kim,

Sorry I was away from keyboard.
You are right, LLVM modules seems to be there but clang's are not :-(

The CMake module seems there (in Ubuntu) since at least since llvm 3.4:
In Ubuntu:
https://packages.ubuntu.com/trusty/amd64/llvm-3.4-dev/filelist
and Debian:
https://packages.debian.org/jessie/amd64/llvm-3.4-dev/filelist
Concerning Windows build I won't be able to help you I don't have access or usage of a Windows box.

Overall I agree, separate llvm's and clang's module finding should help te builder filling the gaps.

**Comment by kimgr** [comment]
Just to follow up on this, I've spent some time trying to get acquainted with the LLVM Debian packaging, and I think the reason I've had so much trouble getting this to work is because I'm hitting https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=862328 when testing on Ubuntu. The FreeBSD packaging seems to be better, and more consistent with the build tree layout.

I'm not sure that this qualifies as a bug in the Debian packaging or just a mere inconsistency, but I haven't been able to pull off `find_package` for Clang yet.

This is something I'd really like to get fixed, so we have a consistent, working build system for IWYU, but it would be nice if didn't have to build in workarounds if packaging is broken. Then again, maybe it's better to do that for now, and work on getting packaging fixed as a lower-prio cleanup task.

**Comment by TheErk** [comment]
Hi Kim,

I'm trying to catch-up on this and I would really like to find a way to build IWYU in-tree with shared LLVM libs enabled. May be this part could be merged in? 
Concerning the clang part with the issues of missing find module in standard distro I guess that asking the user if it cannot be found would be OK for now.

What do you think?
Would you like me to try to backport my current proposal to IWYU master?

**Comment by kimgr** [comment]
Oh, I was just looking at this again last night :)

Yes, I think we want to do something like this -- it should be relatively safe to assume that LLVM `CFLAGS` and other settings are compatible with Clang. So maybe all we need is a way to override Clang include/lib paths if `find_package` for Clang fails?

I'm also wondering whether we really need `llvm-config`? Shouldn't `find_package` be able to provide much of the same information?

**Comment by TheErk** [comment]
Agreed for LLVM/Clang flags compatibilty.

Yes being able to set/provide value for CLANG_INCLUDE_DIRS should be enough.

The use of llvm-config was inspired by clang sources:
https://github.com/llvm-mirror/clang/blob/master/CMakeLists.txt#L10
In the case of out-of-tree build:
https://github.com/llvm-mirror/clang/blob/master/CMakeLists.txt#L5

Beside mimicking Clang source tree it helps to appropriately append
CMAKE_MODULE_PATH for finding FindLLVM and FINdClang CMake module.
If those (or at least LLVM one) could be found without that then there is no need to use llvm-config.

**Comment by kimgr** [comment]
I know preciously little about CMake, but I just used `find_package` with `HINTS` set to the expected location, so I haven't seen how `llvm-config` adds anything. But I'm probably missing some crucial detail -- I'm still learning CMake.

**Comment by TheErk** [comment]
You are right llvm-config is only useful to append CMAKE_MODULE_PATH.
If you provide the same path to find_package HINTS argument then there musn't much benefit of using llvm-config to discover the path to LLVM cmake module.

The main difference CMake-wise is that HINTS only works for find_package (config mode)
i.e. for project which provides <project>Config.cmake.
This is the case of LLVM and Clang (but not Git for example).

That said I may have lost track of your current work with that because current master does not
find_package(..) to anything (neither LLVM nor CLang) am I wrong?
And your branch 're-cmake' has not evolved since last october, or did you forget to push?
May be you did create another branch to work on this I am not aware of?

**Comment by kimgr** [comment]
Thanks for confirming! 

I've only hacked on this on my local machine, and never got anything stable enough to even commit. But I came up with the general approach of preferring `find_package` and provide override variables for when Clang is broken (such as on Ubuntu). 

So no, you haven't lost track -- there was nothing to follow :)

**Comment by TheErk** [comment]
Ok then I'll wait your "public" update :-)

Since you are at it (with find_package) you should replace the current:
include(FindGit) there
https://github.com/include-what-you-use/include-what-you-use/blob/master/CMakeLists.txt#L74
with
find_package(Git)

including a FindXXX module is working but it may fail in the future if the concerned project provides a Config mode file instead of a Module file (see: https://cmake.org/cmake/help/v3.7/command/find_package.html)

moreover 
find_package(Git) 
seems clearer (at least to me) than
include(FindGit)

**Comment by kimgr** [comment]
Messy work-in-progress here: https://github.com/include-what-you-use/include-what-you-use/pull/514

**Comment by TheErk** [comment]
Ok thanks, I follow it there.

**Comment by kimgr** [comment]
OK, I've just about given up on getting this to work for the Debian packaging. There doesn't seem to be a way to consume the Clang CMake package.

On the flip-side, I think I have in-tree builds working fine, and I've seen out-of-tree on FreeBSD work at least once.

I'll take a stab at dylib support per your earlier design and clean out some cruft from Debian workarounds, and maybe you can test in your environment?

Once we have something reasonably stable, I'll try again to get something in place for Debian, as our Travis and buildbot environments both use Ubuntu :-/

**Comment by TheErk** [comment]
I'll definitely give it a try in our environnement.
It'll probably be next week though.

**Comment by TheErk** [comment]
This branch 
https://github.com/TheErk/include-what-you-use/tree/re-cmake-2
seems OK in our environment (not thoroughly tested yet).

You may cherry-pick  or mimics the commit
https://github.com/TheErk/include-what-you-use/commit/c6a92d498e17edd6958a8484cd019a87fea4d5d9

in order to introduce the logic for dylib support.

**Comment by kimgr** [comment]
Quick status update: there does not appear to be a reasonable workaround for Debian systems, so I'm now putting all my fragments of free time into fixing the Debian packaging first. This is necessary to keep our build infra in working order. Sorry this is moving so slow...

**Comment by TheErk** [comment]
Thank you for the heads up.
No problem with the timing I know packaging issue may be painful.

**Comment by TheErk** [comment]
Ok I gave a deeper shot to our llvm+clang+iwyu compilation yesterday and today, here comes the result:

My tests was done using 
llvm/clang from gitlab mirror (https://github.com/llvm-mirror/) on master branch as of 2018-01-18 and iwyu from kimr/re-cmake-2 (https://github.com/kimgr/include-what-you-use)

All compilation of IWYU:

- in-tree (in LLVM/Clang tree) or out-of-tree works: no error and simple command invocation of include-what-you-use command line without arguments (getting help) works

- IWYU build equally well when LLVM is built as a shared lib or not (LLVM_BUILD_LLVM_DYLIB)
  see: https://github.com/include-what-you-use/include-what-you-use/pull/514/commits/9f7cbcded86cb3d36df1c083d86823e8936418b7

There is however one remaining issue. When building in-tree the resulting iwyu command even it seems to run correctly **always** generate a core dump when used in a real case (I cannot provide for now).

When compiling out-of-tree the resulting IWYU command works flawlessly.
From my point of view this issue can be closed when/if kimr/re-cmake-2 is merged.

I'll try to provide a stripped-down core dump example and open a separate issue.

**Comment by kimgr** [comment]
I'll admit that I'm having trouble with the Debian packaging -- I haven't managed to produce a working version yet, and testing is a pain (no doubt because I'm lacking one or two skills).

And because we need Debian packaging working to keep our build infrastructure running with a simplified CMakeLists.txt, we're kind-of stuck.

So let's turn this around -- it seems like originally you had problems when linking LLVM dynamically. Can we see if we can make that scenario work with the current CMakeLists, either implicitly (might be possible for in-tree with a simple conditional link line on `LLVM_BUILD_LLVM_DYLIB`) or explicitly (with an `IWYU_LLVM_DYLIB` Boolean flag)?

I'll see if I have any local LLVM build with a dylib available, and try to hack something up.

**Comment by kimgr** [comment]
@TheErk Could you please give this a shot: https://github.com/include-what-you-use/include-what-you-use/pull/517?

**Comment by TheErk** [comment]
OK with this path.
I'll give a shot to #517 in our context. I'll give you feedback in #517 directly.

**Comment by kimgr** [comment]
This is fixed as of 5550ce2672b3a2a86ef60a77513ea6ac70626a8a. See https://github.com/include-what-you-use/include-what-you-use/blob/master/README.md#how-to-build for updated build instructions.
